### DISCUSSION TITLE
MCU for new controller

### DISCUSSION THREAD

**Discussion by drom89** [grblHAL/Controllers#9] [Q&A]
Hi, we are considering making a new controller, mainly for lasers.
And i want to ask what MCU will be the best for this job.
Im looking for MCU that will be able to handle high speed engraving 1000mm/s +. 

What are the limitations of grbl/grblHAL in this high speeds? 
What should we consider ? 

few MCUs we are looking at: 
i.MX RT from NXP are somewhat available (MIMXRT1062DVJ6B is in teensy)
[https://www.nxp.com/products/processors-and-microcontrollers/arm-microcontrollers/i-mx-rt-crossover-mcus:IMX-RT-SERIES](url)
ST
[https://www.st.com/content/st_com/en/arm-32-bit-microcontrollers.html](url)
TI
[https://www.ti.com/microcontrollers-mcus-processors/c2000-real-time-control-mcus/overview.html](url)

**Comment by terjeio** [comment]
> What are the limitations of grbl/grblHAL in this high speeds?

Max step rate, interrupt latency is the main limiting factor.

> What should we consider ?

Start with evaluating the iMXRT1062 (Teensy 4), this can easily output step rates in excess of what most steppers/stepper drivers can handle (200-250KHz). STM32F7xx or STM32H7xx would be my next choice.

The TI C2000 seems to be limited to 120MHz clock speeds, not what you want since there is quite a bit of processing going on in the stepper interrupt handler.

My guess is that your main challenges will be on the mechanical side - motors able to handle the neccesary step rates/speeds ...

**Reply by drom89** [reply]
What about iMXRT1170 or 1160 it is has dual core Cortex-M7 + Cortex-M4.
Or do you think iMXRT1062 is enough ? For all the possible functionality ? 

I think mechanical side we be ok, 3 phase steppers or servo motors will be used...
Standard stepper motors are possible to handle this speeds too, if the laser head is not too heavy.

**Reply by terjeio** [reply]
> What about iMXRT1170 or 1160 it is has dual core Cortex-M7 + Cortex-M4.

Those would require new drivers? The 1062 uses the Arduino framework, a lot of work to ditch that or port it?

> Or do you think iMXRT1062 is enough ? For all the possible functionality ?

Hard to tell, but it is very capable - if you really need a fast multi core MCU then there is, IMO, a lot of work ahead of you.
